crio static build

we should support a static build for CRI-O so that we don't need to carry dependencies and make it easier to copy the crio binary.

More information here: https://github.com/rootless-containers/usernetes/issues/19

I played around with static cri-o builds and setup a Nix package which produces a 50Mb binary.
https://gist.github.com/saschagrunert/d043c2c8615700b5dae64e967183cc58

Some characteristics:
- go version go1.12 linux/amd64
- binary:
    ```
    > file result-bin/bin/crio
    result-bin/bin/crio: ELF 64-bit LSB executable, x86-64, version 1 (GNU/Linux), statically linked, for GNU/Linux 2.6.32, Go BuildID=wcs6RAOx63B7GLR4ISa6/SuuQZddzDCSqNV1oOKoq/MyRYBEPOHuJdF17Tdc20/JVnN3XJcTYuu_823iLMg, stripped
    ```
- compiler warnings:
    ```
    # net
    /nix/store/0y7jmqnj48ikjh37n3dl9kqw9hnn68nq-binutils-2.31.1/bin/ld: $WORK/b048/_x005.o: in function `_cgo_7e1b3c2abc8d_C2func_getaddrinfo':
    /tmp/go-build/cgo-gcc-prolog:57: warning: Using 'getaddrinfo' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
    # os/user
    /nix/store/0y7jmqnj48ikjh37n3dl9kqw9hnn68nq-binutils-2.31.1/bin/ld: $WORK/b097/_x003.o: in function `mygetgrouplist':
    /nix/store/barxv95b8arrlh97s6axj8k7ljn7aky1-go-1.12/share/go/src/os/user/getgrouplist_unix.go:16: warning: Using 'getgrouplist' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
    /nix/store/0y7jmqnj48ikjh37n3dl9kqw9hnn68nq-binutils-2.31.1/bin/ld: $WORK/b097/_x002.o: in function `mygetgrgid_r':
    /nix/store/barxv95b8arrlh97s6axj8k7ljn7aky1-go-1.12/share/go/src/os/user/cgo_lookup_unix.go:38: warning: Using 'getgrgid_r' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
    /nix/store/0y7jmqnj48ikjh37n3dl9kqw9hnn68nq-binutils-2.31.1/bin/ld: $WORK/b097/_x002.o: in function `mygetgrnam_r':
    /nix/store/barxv95b8arrlh97s6axj8k7ljn7aky1-go-1.12/share/go/src/os/user/cgo_lookup_unix.go:43: warning: Using 'getgrnam_r' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
    /nix/store/0y7jmqnj48ikjh37n3dl9kqw9hnn68nq-binutils-2.31.1/bin/ld: $WORK/b097/_x002.o: in function `mygetpwnam_r':
    /nix/store/barxv95b8arrlh97s6axj8k7ljn7aky1-go-1.12/share/go/src/os/user/cgo_lookup_unix.go:33: warning: Using 'getpwnam_r' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
    /nix/store/0y7jmqnj48ikjh37n3dl9kqw9hnn68nq-binutils-2.31.1/bin/ld: $WORK/b097/_x002.o: in function `mygetpwuid_r':
    /nix/store/barxv95b8arrlh97s6axj8k7ljn7aky1-go-1.12/share/go/src/os/user/cgo_lookup_unix.go:28: warning: Using 'getpwuid_r' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
    # github.com/kubernetes-sigs/cri-o/vendor/github.com/coreos/pkg/dlopen
    /nix/store/0y7jmqnj48ikjh37n3dl9kqw9hnn68nq-binutils-2.31.1/bin/ld: $WORK/b554/_x002.o: in function `_cgo_d4e5ccbb7871_Cfunc_dlopen':
    /tmp/go-build/cgo-gcc-prolog:89: warning: Using 'dlopen' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
    # github.com/kubernetes-sigs/cri-o/cmd/crio
    /nix/store/0y7jmqnj48ikjh37n3dl9kqw9hnn68nq-binutils-2.31.1/bin/ld: /build/go-link-672543141/000044.o: in function `_cgo_d4e5ccbb7871_Cfunc_dlopen':
    /tmp/go-build/cgo-gcc-prolog:89: warning: Using 'dlopen' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
    /nix/store/0y7jmqnj48ikjh37n3dl9kqw9hnn68nq-binutils-2.31.1/bin/ld: /build/go-link-672543141/000027.o: in function `mygetgrouplist':
    /nix/store/barxv95b8arrlh97s6axj8k7ljn7aky1-go-1.12/share/go/src/os/user/getgrouplist_unix.go:16: warning: Using 'getgrouplist' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
    /nix/store/0y7jmqnj48ikjh37n3dl9kqw9hnn68nq-binutils-2.31.1/bin/ld: /build/go-link-672543141/000026.o: in function `mygetgrgid_r':
    /nix/store/barxv95b8arrlh97s6axj8k7ljn7aky1-go-1.12/share/go/src/os/user/cgo_lookup_unix.go:38: warning: Using 'getgrgid_r' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
    /nix/store/0y7jmqnj48ikjh37n3dl9kqw9hnn68nq-binutils-2.31.1/bin/ld: /build/go-link-672543141/000026.o: in function `mygetgrnam_r':
    /nix/store/barxv95b8arrlh97s6axj8k7ljn7aky1-go-1.12/share/go/src/os/user/cgo_lookup_unix.go:43: warning: Using 'getgrnam_r' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
    /nix/store/0y7jmqnj48ikjh37n3dl9kqw9hnn68nq-binutils-2.31.1/bin/ld: /build/go-link-672543141/000026.o: in function `mygetpwnam_r':
    /nix/store/barxv95b8arrlh97s6axj8k7ljn7aky1-go-1.12/share/go/src/os/user/cgo_lookup_unix.go:33: warning: Using 'getpwnam_r' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
    /nix/store/0y7jmqnj48ikjh37n3dl9kqw9hnn68nq-binutils-2.31.1/bin/ld: /build/go-link-672543141/000026.o: in function `mygetpwuid_r':
    /nix/store/barxv95b8arrlh97s6axj8k7ljn7aky1-go-1.12/share/go/src/os/user/cgo_lookup_unix.go:28: warning: Using 'getpwuid_r' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
    /nix/store/0y7jmqnj48ikjh37n3dl9kqw9hnn68nq-binutils-2.31.1/bin/ld: /build/go-link-672543141/000004.o: in function `_cgo_7e1b3c2abc8d_C2func_getaddrinfo':
    /tmp/go-build/cgo-gcc-prolog:57: warning: Using 'getaddrinfo' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
    ```

The binary works (did only some basic tests with `crictl`), whereas the following things are not done yet:
- lvm2 static build for libdevmapper
- gpgme static build containers/image
- ostree static build

I also tried to build it on top of the musl toolchain, but this leads into compile time errors for dependencies like go and systemd, so I see a huge effort there.

So *one* appraoch could be to build a NixOS container image which is [based on nixos/nix](https://hub.docker.com/r/nixos/nix). We could prefetch/pin the Nix packages to increase the build time and have a full reproducable build. Then it should be fairly easy to pass another CRI-O version via an env variable in. WDYT?

May I ask you which targets/architectures we should support?

--- a/.gitignore
+++ b/.gitignore
@@ -18,6 +18,7 @@ coverprofile
 .gdb_history
 *gomock_reflect*
 lib/state.json
+result-*bin
 
 Vagrantfile
 .vagrant/

--- a/.travis.yml
+++ b/.travis.yml
@@ -7,8 +7,6 @@ services:
   - docker
 
 before_install:
-  - if [ "${TRAVIS_OS_NAME}" = linux ]; then make vendor; fi
-  - if [ "${TRAVIS_OS_NAME}" = linux ]; then ./hack/tree_status.sh; fi
   - if [ "${TRAVIS_OS_NAME}" = linux ]; then sudo apt-get -qq update; fi
   - if [ "${TRAVIS_OS_NAME}" = linux ]; then sudo apt-get -qq install btrfs-tools libdevmapper-dev libgpgme11-dev libseccomp-dev; fi
   - if [ "${TRAVIS_OS_NAME}" = linux ]; then sudo apt-get -qq install autoconf automake bison clang-format-3.9 e2fslibs-dev libfuse-dev libtool liblzma-dev gettext libsystemd-dev; fi
@@ -20,40 +18,70 @@ before_script:
   - export LD_LIBRARY_PATH=/usr/local/lib${LD_LIBRARY_PATH:+:${LD_LIBRARY_PATH}}
 
 env:
+  global:
+    - NIX_IMAGE=saschagrunert/crionix
+    - CONTAINER_RUNTIME=docker
 
 jobs:
   include:
-    - stage: Build and Verify
+    - stage: Lint
+      name: git validation, clang-format, golangci-lint
       script:
         - make .gitvalidation
         - make fmt
         - make lint
       go: "1.12.x"
-    - script:
+    - name: vendor
+      script:
+        - make vendor
+        - hack/tree_status.sh
+      go: "1.12.x"
+    - stage: Test
+      name: build, unit test and code coverage report
+      script:
         - make testunit
         - make
         - make codecov
       go: "1.12.x"
-    - script:
+    - name: static binary build
+      script:
+        - make build-static CONTAINER_RUNTIME=$CONTAINER_RUNTIME NIX_IMAGE=$NIX_IMAGE
+      go: "1.12.x"
+    - name: cross compilation build
+      script:
         - make local-cross
       go: "1.12.x"
       env: CROSS_PLATFORM=true
-    - script:
+    - name: latest go version (tip)
+      script:
         - make testunit
         - make
       go: tip
-    - script:
+    - name: macOS build and unit tests
+      script:
         - make testunit
         - make
       os: osx
-    - stage: Integration Test
+    - stage: Integration
+      name: default build
       script:
-        - make CONTAINER_RUNTIME=docker integration
+        - make integration CONTAINER_RUNTIME=$CONTAINER_RUNTIME
       go: "1.12.x"
-    - script:
-        - make CONTAINER_RUNTIME=docker integration
+    - name: default build with user namespace
+      script:
+        - make integration CONTAINER_RUNTIME=$CONTAINER_RUNTIME
       env: TEST_USERNS=1
       go: "1.12.x"
+    - name: static x86_64 glibc
+      script:
+        - make build-static integration CONTAINER_RUNTIME=$CONTAINER_RUNTIME NIX_IMAGE=$NIX_IMAGE
+      go: "1.12.x"
+      env: CRIO_BINARY=crio-x86_64-static-glibc
+    - name: static x86_64 musl libc
+      script:
+        - make build-static integration CONTAINER_RUNTIME=$CONTAINER_RUNTIME NIX_IMAGE=$NIX_IMAGE
+      go: "1.12.x"
+      env: CRIO_BINARY=crio-x86_64-static-musl
   allow_failures:
     - os: osx
     - go: tip

--- a/Dockerfile-nix
+++ b/Dockerfile-nix
@@ -1,0 +1,8 @@
+FROM nixos/nix
+RUN apk add git
+COPY . /cri-o
+ARG COMMIT
+WORKDIR cri-o/nix
+RUN nix-build --argstr revision ${COMMIT}
+WORKDIR /
+RUN rm -rf cri-o

--- a/Makefile
+++ b/Makefile
@@ -47,6 +47,8 @@ GIT_VALIDATION := ${BUILD_BIN_PATH}/git-validation
 RELEASE_TOOL := ${BUILD_BIN_PATH}/release-tool
 GOLANGCI_LINT := ${BUILD_BIN_PATH}/golangci-lint
 
+NIX_IMAGE := crionix
+
 CROSS_BUILD_TARGETS := \
 	bin/crio.cross.windows.amd64 \
 	bin/crio.cross.darwin.amd64 \
@@ -117,6 +119,17 @@ test/checkseccomp/checkseccomp: .gopathok $(wildcard test/checkseccomp/*.go)
 bin/crio: .gopathok
 	$(GO) build -i $(LDFLAGS) -tags "$(BUILDTAGS)" -o $@ $(PROJECT)/cmd/crio
 
+build-static:
+	$(CONTAINER_RUNTIME) run --rm -it -v $(shell pwd):/cri-o $(NIX_IMAGE) sh -c \
+		"nix-build cri-o/nix --argstr revision $(COMMIT_NO) && \
+		mkdir -p cri-o/bin && \
+		cp result-*bin/bin/crio-* cri-o/bin && \
+		chown -R $(shell id -u):$(shell id -g) cri-o/bin"
+
+nix-image:
+	time $(CONTAINER_RUNTIME) build -t $(NIX_IMAGE) \
+		--build-arg COMMIT=$(COMMIT_NO) -f Dockerfile-nix .
+
 crio.conf: bin/crio
 	./bin/crio --config="" config --default > crio.conf
 
@@ -165,7 +178,7 @@ dbuild: crioimage
 	$(CONTAINER_RUNTIME) run --name=${CRIO_INSTANCE} -e BUILDTAGS --privileged -v ${PWD}:/go/src/${PROJECT} --rm ${CRIO_IMAGE} make binaries
 
 integration: crioimage
-	$(CONTAINER_RUNTIME) run -e STORAGE_OPTIONS="--storage-driver=vfs" -e TEST_USERNS -e TESTFLAGS -e TRAVIS -t --privileged --rm -v ${CURDIR}:/go/src/${PROJECT} ${CRIO_IMAGE} make localintegration
+	$(CONTAINER_RUNTIME) run -e STORAGE_OPTIONS="--storage-driver=vfs" -e TEST_USERNS -e TESTFLAGS -e TRAVIS -e CRIO_BINARY -t --privileged --rm -v ${CURDIR}:/go/src/${PROJECT} ${CRIO_IMAGE} make localintegration
 
 define go-build
 	$(shell cd `pwd` && $(GO) build -o ${BUILD_BIN_PATH}/${1} ${2})
@@ -327,6 +340,7 @@ endif
 	bin/crio \
 	bin/pause \
 	binaries \
+	build-static \
 	clean \
 	default \
 	docs \
@@ -334,5 +348,6 @@ endif
 	install \
 	lint \
 	local-cross \
+	nix-image \
 	uninstall \
 	vendor

--- a/nix/default.nix
+++ b/nix/default.nix
@@ -1,0 +1,88 @@
+{ revision ? "HEAD", system ? builtins.currentSystem }:
+let
+  nixpkgs = import ./nixpkgs.nix;
+
+  glibcPkgs = import (builtins.head nixpkgs) {
+    config = { packageOverrides = pkg: { go_1_11 = glibcPkgs.go_1_12; }; };
+  };
+  glibcCallPackage = glibcPkgs.lib.callPackageWith
+                     (glibcPkgs // glibcPkgs.xlibs // self);
+
+  muslPkgs = (import (builtins.head nixpkgs) {
+    config = {
+      packageOverrides = pkg: {
+        go_bootstrap = pkg.go_bootstrap.overrideAttrs (old: {
+          installPhase = ''
+            sed -i 's/TestChown/testChown/' src/os/os_unix_test.go
+          '' + old.installPhase;
+        });
+        go_1_11 = muslPkgs.go_1_12;
+        go_1_12 = pkg.go_1_12.overrideAttrs (old: {
+          prePatch = ''
+            sed -i 's/TestChown/testChown/' src/os/os_unix_test.go
+            sed -i 's/TestFileChown/testFileChown/' src/os/os_unix_test.go
+            sed -i 's/TestLchown/testLchown/' src/os/os_unix_test.go
+            sed -i 's/TestCoverageUsesAtomicModeForRace/testCoverageUsesAtomicModeForRace/' src/cmd/go/go_test.go
+            sed -i 's/TestTestRaceInstall/testTestRaceInstall/' src/cmd/go/go_test.go
+            sed -i 's/TestGoTestRaceFailures/testGoTestRaceFailures/' src/cmd/go/go_test.go
+            sed -i '/func cmdtest/a return' src/cmd/dist/test.go
+          '' + old.prePatch;
+        });
+        gnupg = glibcPkgs.gnupg;
+        systemd = (import (builtins.elemAt nixpkgs 1) {}).systemd;
+      };
+    };
+  }).pkgsMusl;
+  muslCallPackage = muslPkgs.lib.callPackageWith
+                    (muslPkgs // muslPkgs.xlibs // self);
+
+  static = pkg: pkg.overrideAttrs(old: {
+    configureFlags = (old.configureFlags or []) ++
+      [ "--without-shared" "--disable-shared" ];
+    dontDisableStatic = true;
+    enableSharedExecutables = false;
+    enableStatic = true;
+  });
+
+  patchLvm2 = pkg: pkg.overrideAttrs(old: {
+    configureFlags = [
+      "--disable-cmdlib" "--disable-readline" "--disable-udev_rules"
+      "--disable-udev_sync" "--enable-pkgconfig" "--enable-static_link"
+    ];
+    preConfigure = old.preConfigure + ''
+      substituteInPlace libdm/Makefile.in --replace \
+        SUBDIRS=dm-tools SUBDIRS=
+      substituteInPlace tools/Makefile.in --replace \
+        "TARGETS += lvm.static" ""
+      substituteInPlace tools/Makefile.in --replace \
+        "INSTALL_LVM_TARGETS += install_tools_static" ""
+    '';
+    postInstall = "";
+  });
+
+  self = {
+    cri-o-static-musl = muslCallPackage ./derivation.nix {
+      flavor = "x86_64-static-musl";
+      ldflags = ''-linkmode external -extldflags "-static"'';
+      revision = revision;
+
+      glibc = null;
+      gpgme = (static muslPkgs.gpgme);
+      libassuan = (static muslPkgs.libassuan);
+      libgpgerror = (static muslPkgs.libgpgerror);
+      libseccomp = (static muslPkgs.libseccomp);
+      lvm2 = (patchLvm2 (static muslPkgs.lvm2));
+    };
+    cri-o-static-glibc = glibcCallPackage ./derivation.nix {
+      flavor = "x86_64-static-glibc";
+      ldflags = ''-linkmode external -extldflags "-static -lm"'';
+      revision = revision;
+
+      gpgme = (static glibcPkgs.gpgme);
+      libassuan = (static glibcPkgs.libassuan);
+      libgpgerror = (static glibcPkgs.libgpgerror);
+      libseccomp = (static glibcPkgs.libseccomp);
+      lvm2 = (patchLvm2 (static glibcPkgs.lvm2));
+    };
+  };
+in self

--- a/nix/derivation.nix
+++ b/nix/derivation.nix
@@ -1,0 +1,34 @@
+{ flavor, ldflags ? "", revision,
+  btrfs-progs, buildGoPackage, glibc, gpgme, libapparmor, libassuan,
+  libgpgerror, libseccomp, libselinux, lvm2, pkgconfig, stdenv }:
+
+buildGoPackage rec {
+  project = "cri-o";
+  name = "${project}-${revision}-${flavor}";
+
+  goPackagePath = "github.com/${project}/${project}";
+  src = ./..;
+
+  outputs = [ "bin" "out" ];
+  nativeBuildInputs = [ pkgconfig ];
+  buildInputs = [ btrfs-progs gpgme libapparmor libassuan
+                  libgpgerror libseccomp libselinux lvm2 ] ++
+                stdenv.lib.optionals (glibc != null) [ glibc glibc.static ];
+
+  makeFlags = ''BUILDTAGS="apparmor seccomp selinux
+    containers_image_ostree_stub"'';
+
+  buildPhase = ''
+    go build -tags ${makeFlags} -o bin/crio -buildmode=pie \
+      -ldflags '-s -w ${ldflags}' ${goPackagePath}/cmd/crio
+  '';
+  installPhase = "install -Dm755 bin/crio $bin/bin/crio-${flavor}";
+
+  meta = with import <nixpkgs/lib>; {
+    homepage = https://cri-o.io;
+    description = ''Open Container Initiative-based implementation of
+                    the Kubernetes Container Runtime Interface'';
+    license = licenses.asl20;
+    platforms = platforms.linux;
+  };
+}

--- a/nix/nixpkgs.nix
+++ b/nix/nixpkgs.nix
@@ -1,0 +1,14 @@
+let
+  nixpkgs = builtins.fetchTarball {
+    name = "nixos-unstable";
+    url = "https://github.com/nixos/nixpkgs/archive/" +
+      "03d6bb52b250d74d096e2286b02c51f595fa90ee.tar.gz";
+    sha256 = "0bngjc4d190bgzhgxv9jszqkapx3ngdpj08rd2arz19ncggflnxv";
+  };
+  nixpkgsMuslSystemd = builtins.fetchTarball {
+    name = "nixos-systemd-musl";
+    url = "https://github.com/dtzWill/nixpkgs/archive/" +
+      "13510d3dbe08e5bfc5454faf3fe543991dbf6e29.tar.gz";
+    sha256 = "090zagbw39fa8fgwvzcmbcr204pxcfrq1rfw7ip5z44naj6gp7qb";
+  };
+in [ nixpkgs nixpkgsMuslSystemd ]

--- a/test/command.bats
+++ b/test/command.bats
@@ -3,20 +3,20 @@
 load helpers
 
 @test "crio commands" {
-	run ${CRIO_BINARY} --config /dev/null config > /dev/null
+	run ${CRIO_BINARY_PATH} --config /dev/null config > /dev/null
 	echo "$output"
 	[ "$status" -eq 0 ]
-	run ${CRIO_BINARY} badoption > /dev/null
+	run ${CRIO_BINARY_PATH} badoption > /dev/null
 	echo "$output"
 	[ "$status" -ne 0 ]
 }
 
 @test "invalid ulimits" {
-	run ${CRIO_BINARY} --default-ulimits doesntexist=2042
+	run ${CRIO_BINARY_PATH} --default-ulimits doesntexist=2042
 	echo $output
 	[ "$status" -ne 0 ]
 	[[ "$output" =~ "invalid ulimit type: doesntexist" ]]
-	run ${CRIO_BINARY} --default-ulimits nproc=2042:42
+	run ${CRIO_BINARY_PATH} --default-ulimits nproc=2042:42
 	echo $output
 	[ "$status" -ne 0 ]
 	[[ "$output" =~ "ulimit soft limit must be less than or equal to hard limit: 2042 > 42" ]]
@@ -25,26 +25,26 @@ load helpers
 }
 
 @test "invalid devices" {
-	run ${CRIO_BINARY} --additional-devices /dev/sda:/dev/foo:123
+	run ${CRIO_BINARY_PATH} --additional-devices /dev/sda:/dev/foo:123
 	echo $output
 	[ "$status" -ne 0 ]
 	[[ "$output" =~ "invalid device mode:" ]]
-	run ${CRIO_BINARY} --additional-devices /dev/sda:/dee/foo:rm
+	run ${CRIO_BINARY_PATH} --additional-devices /dev/sda:/dee/foo:rm
 	echo $output
 	[ "$status" -ne 0 ]
 	[[ "$output" =~ "invalid device mode:" ]]
-	run ${CRIO_BINARY} --additional-devices /dee/sda:rmw
+	run ${CRIO_BINARY_PATH} --additional-devices /dee/sda:rmw
 	echo $output
 	[ "$status" -ne 0 ]
 	[[ "$output" =~ "invalid device mode:" ]]
 }
 
 @test "invalid metrics port" {
-	run ${CRIO_BINARY} --metrics-port foo --enable-metrics
+	run ${CRIO_BINARY_PATH} --metrics-port foo --enable-metrics
 	echo $output
 	[ "$status" -ne 0 ]
 	[[ "$output" =~ "invalid value \"foo\" for flag" ]]
-	run ${CRIO_BINARY} --metrics-port 18446744073709551616 --enable-metrics
+	run ${CRIO_BINARY_PATH} --metrics-port 18446744073709551616 --enable-metrics
 	echo $output
 	[ "$status" -ne 0 ]
 	[[ "$output" =~ "value out of range" ]]

--- a/test/helpers.bash
+++ b/test/helpers.bash
@@ -9,8 +9,11 @@ TESTDATA="${INTEGRATION_ROOT}/testdata"
 # Root directory of the repository.
 CRIO_ROOT=${CRIO_ROOT:-$(cd "$INTEGRATION_ROOT/../.."; pwd -P)}
 
-# Path of the crio binary.
-CRIO_BINARY=${CRIO_BINARY:-${CRIO_ROOT}/cri-o/bin/crio}
+# Path to the crio binary.
+CRIO_BINARY=${CRIO_BINARY:-crio}
+CRIO_BINARY_PATH=${CRIO_BINARY_PATH:-${CRIO_ROOT}/cri-o/bin/$CRIO_BINARY}
+echo "Using cri-o binary '$CRIO_BINARY_PATH'"
+
 # Path of the crictl binary.
 CRICTL_PATH=$(command -v crictl || true)
 CRICTL_BINARY=${CRICTL_PATH:-/usr/bin/crictl}
@@ -184,10 +187,10 @@ if ! [ -d "$ARTIFACTS_PATH"/image-volume-test-image ]; then
         exit 1
     fi
 fi
-# Run crio using the binary specified by $CRIO_BINARY.
+# Run crio using the binary specified by $CRIO_BINARY_PATH.
 # This must ONLY be run on engines created with `start_crio`.
 function crio() {
-	"$CRIO_BINARY" --listen "$CRIO_SOCKET" "$@"
+	"$CRIO_BINARY_PATH" --listen "$CRIO_SOCKET" "$@"
 }
 
 # Run crictl using the binary specified by $CRICTL_BINARY.
@@ -255,7 +258,7 @@ function setup_crio() {
 	"$COPYIMG_BINARY" --root "$TESTDIR/crio" $STORAGE_OPTIONS --runroot "$TESTDIR/crio-run" --image-name=quay.io/crio/image-volume-test:latest --import-from=dir:"$ARTIFACTS_PATH"/image-volume-test-image --signature-policy="$INTEGRATION_ROOT"/policy.json
 	"$COPYIMG_BINARY" --root "$TESTDIR/crio" $STORAGE_OPTIONS --runroot "$TESTDIR/crio-run" --image-name=quay.io/crio/busybox:latest --import-from=dir:"$ARTIFACTS_PATH"/busybox-image --signature-policy="$INTEGRATION_ROOT"/policy.json
 	"$COPYIMG_BINARY" --root "$TESTDIR/crio" $STORAGE_OPTIONS --runroot "$TESTDIR/crio-run" --image-name=quay.io/crio/stderr-test:latest --import-from=dir:"$ARTIFACTS_PATH"/stderr-test --signature-policy="$INTEGRATION_ROOT"/policy.json
-	"$CRIO_BINARY" ${DEFAULT_MOUNTS_OPTS} ${HOOKS_OPTS} --default-capabilities "$capabilities" --conmon "$TESTDIR/conmon" --listen "$CRIO_SOCKET" --stream-port "$((STREAM_PORT + BATS_TEST_NUMBER))" --cgroup-manager "$CGROUP_MANAGER" --default-mounts-file "$TESTDIR/containers/mounts.conf" --registry "quay.io" --default-runtime $DEFAULT_RUNTIME --runtimes "$RUNTIME_NAME:$RUNTIME_BINARY:$RUNTIME_ROOT" --root "$TESTDIR/crio" --runroot "$TESTDIR/crio-run" $STORAGE_OPTIONS --seccomp-profile "$seccomp" --apparmor-profile "$apparmor" --cni-config-dir "$CRIO_CNI_CONFIG" --cni-plugin-dir "$CRIO_CNI_PLUGIN" --signature-policy "$INTEGRATION_ROOT"/policy.json --image-volumes "$IMAGE_VOLUMES" --pids-limit "$PIDS_LIMIT" --log-size-max "$LOG_SIZE_MAX_LIMIT" $DEVICES $ULIMITS --uid-mappings "$UID_MAPPINGS" --gid-mappings "$GID_MAPPINGS" --default-sysctls "$TEST_SYSCTL" $OVERRIDE_OPTIONS --config /dev/null config >$CRIO_CONFIG
+	"$CRIO_BINARY_PATH" ${DEFAULT_MOUNTS_OPTS} ${HOOKS_OPTS} --default-capabilities "$capabilities" --conmon "$TESTDIR/conmon" --listen "$CRIO_SOCKET" --stream-port "$((STREAM_PORT + BATS_TEST_NUMBER))" --cgroup-manager "$CGROUP_MANAGER" --default-mounts-file "$TESTDIR/containers/mounts.conf" --registry "quay.io" --default-runtime $DEFAULT_RUNTIME --runtimes "$RUNTIME_NAME:$RUNTIME_BINARY:$RUNTIME_ROOT" --root "$TESTDIR/crio" --runroot "$TESTDIR/crio-run" $STORAGE_OPTIONS --seccomp-profile "$seccomp" --apparmor-profile "$apparmor" --cni-config-dir "$CRIO_CNI_CONFIG" --cni-plugin-dir "$CRIO_CNI_PLUGIN" --signature-policy "$INTEGRATION_ROOT"/policy.json --image-volumes "$IMAGE_VOLUMES" --pids-limit "$PIDS_LIMIT" --log-size-max "$LOG_SIZE_MAX_LIMIT" $DEVICES $ULIMITS --uid-mappings "$UID_MAPPINGS" --gid-mappings "$GID_MAPPINGS" --default-sysctls "$TEST_SYSCTL" $OVERRIDE_OPTIONS --config /dev/null config >$CRIO_CONFIG
 	sed -r -e 's/^(#)?root =/root =/g' -e 's/^(#)?runroot =/runroot =/g' -e 's/^(#)?storage_driver =/storage_driver =/g' -e '/^(#)?storage_option = (\[)?[ \t]*$/,/^#?$/s/^(#)?//g' -e '/^(#)?registries = (\[)?[ \t]*$/,/^#?$/s/^(#)?//g' -e '/^(#)?default_ulimits = (\[)?[ \t]*$/,/^#?$/s/^(#)?//g' -i $CRIO_CONFIG
 	sed -ie 's;\(container_exits_dir =\) \(.*\);\1 "'$CONTAINER_EXITS_DIR'";g' $CRIO_CONFIG
 	sed -ie 's;\(container_attach_socket_dir =\) \(.*\);\1 "'$CONTAINER_ATTACH_SOCKET_DIR'";g' $CRIO_CONFIG
@@ -301,7 +304,7 @@ function pull_test_containers() {
 # Start crio.
 function start_crio() {
 	setup_crio "$@"
-	"$CRIO_BINARY" --default-mounts-file "$TESTDIR/containers/mounts.conf" --log-level debug --config "$CRIO_CONFIG" & CRIO_PID=$!
+	"$CRIO_BINARY_PATH" --default-mounts-file "$TESTDIR/containers/mounts.conf" --log-level debug --config "$CRIO_CONFIG" & CRIO_PID=$!
 	wait_until_reachable
 	pull_test_containers
 }
@@ -309,7 +312,7 @@ function start_crio() {
 # Start crio with journald logging
 function start_crio_journald() {
 	setup_crio "$@"
-	"$CRIO_BINARY" --default-mounts-file "$TESTDIR/containers/mounts.conf" --log-level debug --log-journald --config "$CRIO_CONFIG" & CRIO_PID=$!
+	"$CRIO_BINARY_PATH" --default-mounts-file "$TESTDIR/containers/mounts.conf" --log-level debug --log-journald --config "$CRIO_CONFIG" & CRIO_PID=$!
 	wait_until_reachable
 	pull_test_containers
 }
@@ -332,7 +335,7 @@ function check_journald() {
 # Start crio with metrics enable
 function start_crio_metrics() {
 	setup_crio "$@"
-	"$CRIO_BINARY" --default-mounts-file "$TESTDIR/containers/mounts.conf" --log-level debug --metrics-port "$METRICS_PORT" --enable-metrics --config "$CRIO_CONFIG" & CRIO_PID=$!
+	"$CRIO_BINARY_PATH" --default-mounts-file "$TESTDIR/containers/mounts.conf" --log-level debug --metrics-port "$METRICS_PORT" --enable-metrics --config "$CRIO_CONFIG" & CRIO_PID=$!
 	wait_until_reachable
 	pull_test_containers
 }

